### PR TITLE
Deduplicate discovery certificates and add scheduled orphan cleanup

### PR DESCRIPTION
**Pull request by 3keyroman**
Discovery now stores certificate content once in a shared `certificate` table keyed by the SHA-256 fingerprint of the base64 content, with `discovery_certificate` reduced to a join that keeps its own `uuid` and `meta`. A scheduled background job periodically removes certificates that are no longer referenced by any discovery.

- New `certificate` and `certificate_cleanup_state` tables; `discovery_certificate` gains a `certificate` foreign key and its `discovery_id` becomes a real foreign key to `discovery_history`.
- Forward-only data migration (0002) deduplicates existing content, backfills the references, drops rows with a dangling discovery, and removes the old `base64content` column. Existing deployments upgrade transparently with no data loss and no manual steps.
- Discovery persistence performs WinRM collection outside the database transaction and takes a shared advisory lock; the cleanup sweep takes the matching exclusive lock and runs at most once per interval across workers.
- Discovery and authority API responses are unchanged.
- Cleanup is configurable via `CERTIFICATE_CLEANUP_ENABLED` and `CERTIFICATE_CLEANUP_INTERVAL_SECONDS`.
- Requires PostgreSQL 14+.

### CODE REVIEW

File: pyadcs_connector/PyADCSConnector/migrations/0002_shared_certificate_store.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+                     row_number() OVER (PARTITION BY {FP} ORDER BY id) rn
+              FROM "{S}"."discovery_certificate") s WHERE rn = 1;
+
+            UPDATE "{S}"."discovery_certificate" dc
+              SET certificate_id = c.id FROM "{S}"."certificate" c
+              WHERE c.fingerprint = {FP_DC};
+
+            DELETE FROM "{S}"."discovery_certificate" dc
+              WHERE NOT EXISTS (SELECT 1 FROM "{S}"."discovery_history" dh WHERE dh.id = dc.discovery_id);
+            """,
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply pyadcs_connector/PyADCSConnector/migrations/0002_shared_certificate_store.py:90]: Addressed in 7e968a8. The backfill INSERT now sources only rows whose discovery still exists (`WHERE EXISTS (SELECT 1 FROM discovery_history dh WHERE dh.id = dc.discovery_id)`), so the upgrade no longer creates certificate rows for dangling-only content that would be immediately orphaned. Every surviving (non-dangling) row still gets a matching certificate before the `NOT NULL` step. Also added a migration-test assertion that dangling-only content yields zero certificates and the total equals exactly the four still-referenced contents.

File: pyadcs_connector/pyadcs_connector/wsgi.py [hunk 2]
```diff
@@ -14,3 +14,9 @@
 os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'pyadcs_connector.settings')
 
 application = get_wsgi_application()
+
+# Started here (not from AppConfig.ready()) so the scheduler never runs during
+# `manage.py migrate` / `manage.py test`, only when the app is actually served.
+from PyADCSConnector.services.certificate_cleanup import start_cleanup_scheduler
+
+start_cleanup_scheduler()
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply pyadcs_connector/pyadcs_connector/wsgi.py]: Addressed in 7e968a8. `start_cleanup_scheduler()` is now also called from `asgi.py` (mirroring `wsgi.py`), so the sweep starts under both server entrypoints. Starting from both is safe — the function is guarded (`_started`/`_lock`) to start at most once per process — and neither entrypoint is imported by `manage.py migrate/test/check`, so it still never runs during those commands. (The shipped image serves via gunicorn WSGI in `entry.sh`.)

File: pyadcs_connector/PyADCSConnector/services/certificate_cleanup.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            if state and state.last_run_at and (now - state.last_run_at) < interval:
+                return None  # not due yet
+
+            cur.execute(
+                sql.SQL(
+                    "DELETE FROM {schema}.{certificate} c "
+                    "WHERE NOT EXISTS ("
+                    "SELECT 1 FROM {schema}.{discovery_certificate} dc "
+                    "WHERE dc.certificate_id = c.id)"
+                ).format(
```
Comment (lubomirw) [inline pyadcs_connector/PyADCSConnector/services/certificate_cleanup.py:80]: **[Medium]** | Performance

The orphan sweep deletes **all** orphaned certificates in one statement while holding the exclusive transaction advisory lock. Concurrent discovery persistence (shared lock) blocks for the entire delete — a large backlog (e.g. the first run after migration) can stall discovery writes.

_Suggested:_ delete in bounded batches (`... WHERE id IN (SELECT ... LIMIT N)`) across separate short transactions so the shared lock is released between batches.

<sub>flagged by superpowers code-review</sub>

File: pyadcs_connector/PyADCSConnector/tests_discovery.py [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+
+        discovery_history.refresh_from_db()
+        self.assertEqual(discovery_history.status, DiscoveryStatus.FAILED.value)
+        self.assertEqual(len(discovery_history.meta), 1)
+        self.assertEqual(discovery_history.meta[0]["content"][0]["data"], str(boom))
+
+
+class CertificateCleanupTests(TransactionTestCase):
+    """
+    Exercises the Task 3 scheduled orphan sweep (run_cleanup_once): the
```
Comment (lubomirw) [inline pyadcs_connector/PyADCSConnector/tests_discovery.py]: **[Medium]** | Testing

The core safety property — the sweep must skip while a concurrent discovery holds the shared advisory lock — has no test, and the `run_cleanup_once()` returns-None-when-lock-held path is never exercised.

_Suggested:_ hold `pg_advisory_xact_lock_shared(LOCK_KEY)` on a second DB connection, assert `run_cleanup_once()` returns None and deletes nothing, then release and assert it sweeps.

<sub>flagged by superpowers code-review</sub>

File: pyadcs_connector/PyADCSConnector/services/certificate_cleanup.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            if not cur.fetchone()[0]:
+                return None  # another worker holds it
+
+            state = CertificateCleanupState.objects.select_for_update().first()
+
+            cur.execute("SELECT now()")
+            now = cur.fetchone()[0]
+
+            if state and state.last_run_at and (now - state.last_run_at) < interval:
+                return None  # not due yet
```
Comment (lubomirw) [inline pyadcs_connector/PyADCSConnector/services/certificate_cleanup.py:46]: **[Low]** | Correctness

`CertificateCleanupState` has no singleton enforcement: `.first()` has no `order_by` and a new row is created when none exists. Migration 0002 seeds exactly one row so this is fine today, but if a second row ever appears, `.first()` returns an arbitrary row on Postgres and the lease/interval gate becomes nondeterministic.

_Suggested:_ pin the row to a fixed pk (`get_or_create(pk=1)`) or add a single-row constraint, and operate on that fixed row.

<sub>flagged by superpowers + guided-pr-review (2 sources)</sub>

File: pyadcs_connector/PyADCSConnector/utils/certificate_fingerprint.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import hashlib
+
+
+def certificate_fingerprint(base64content: str) -> str:
```
Comment (lubomirw) [inline pyadcs_connector/PyADCSConnector/utils/certificate_fingerprint.py:4]: **[Low]** | Correctness

The fingerprint hashes the exact base64 string (deliberately, for parity with migration 0002 SQL). Two byte-different encodings of the same DER certificate — line wrapping, trailing whitespace — produce different fingerprints and separate `Certificate` rows, so dedup effectiveness silently depends on ADCS returning byte-identical base64.

_Suggested:_ if cross-encoding dedup is wanted, base64-decode to DER and hash the raw bytes; otherwise state this as an explicit accepted limitation in the docstring/README.

<sub>flagged by superpowers code-review</sub>

File: pyadcs_connector/pyadcs_connector/wsgi.py [hunk 7]
```diff
@@ -14,3 +14,11 @@
...
 
 application = get_wsgi_application()
+
+# Started from the WSGI/ASGI entrypoints (see also asgi.py), not from
+# AppConfig.ready(), so the scheduler runs only when the app is actually served --
+# never during `manage.py migrate` / `manage.py test` / `manage.py check`. Starting
+# it from both entrypoints is safe: start_cleanup_scheduler() starts at most once.
+from PyADCSConnector.services.certificate_cleanup import start_cleanup_scheduler
+
+start_cleanup_scheduler()
```
Comment (lubomirw) [inline pyadcs_connector/pyadcs_connector/wsgi.py:34]: **[Low]** | Maintainability

Two edge cases beyond the surrounding comment: (a) `manage.py runserver` imports the WSGI app, so a developer's DB silently gets background orphan DELETEs; (b) under gunicorn `--preload` the thread starts in the pre-fork master and does not survive `fork()` into workers, so the sweep silently stops with no error. Current `entry.sh` has no `--preload`, so (b) is latent.

_Suggested:_ start the scheduler from a gunicorn `post_fork` hook (or an AppConfig.ready guarded to server processes), or document the `--preload` incompatibility next to the entrypoint call.

<sub>flagged by superpowers + guided-pr-review (2 sources)</sub>

File: pyadcs_connector/PyADCSConnector/tests_discovery.py [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+    run_discovery,
+)
+from PyADCSConnector.utils.certificate_fingerprint import certificate_fingerprint
+from PyADCSConnector.utils.discovery_status import DiscoveryStatus
+from PyADCSConnector.utils.dump_parser import DumpParser, ParseResult
+
+
+class DiscoveryDeduplicationTests(TransactionTestCase):
+    """
+    Exercises the post-Task-1 discover_certificates persistence (shared
```
Comment (lubomirw) [inline pyadcs_connector/PyADCSConnector/tests_discovery.py]: **[Nit]** | Maintainability

Test docstrings leak internal implementation-plan task numbers (also at lines 217, 236, 346: 'post-Task-1', '-- Task 3', 'Task 2 now computes', 'the Task 3 scheduled orphan sweep') that mean nothing to a reader of the repo.

_Suggested:_ describe the behavior directly and drop the task numbers.

<sub>flagged by hygiene scan</sub>

**Review by lubomirw — COMMENTED** [review]
Automated multi-reviewer triage — 6 findings (COMMENT). Sources: Copilot CLI (clear), guided-pr-review, superpowers code-review, hygiene scan. No Critical/High; the two Mediums (unbounded delete under lock, missing concurrency test) are the most worth addressing before merge.

File: pyadcs_connector/PyADCSConnector/services/certificate_cleanup.py (hunk 3 above)
Comment (3keyroman) [reply pyadcs_connector/PyADCSConnector/services/certificate_cleanup.py:80]: Addressed in 5400a11. The sweep now deletes in bounded batches: `run_cleanup_once()` claims the lease in one short transaction, then each batch re-takes the exclusive lock in its own transaction and deletes at most `CERTIFICATE_CLEANUP_BATCH_SIZE` rows (new setting, default 1000), releasing the lock on commit. Concurrent discovery persistence therefore waits at most one batch instead of the whole sweep. Each sweep is also bounded to the certificates that existed when it claimed the lease (`c.id <= max(id)` captured under the lock), so continuous churn cannot extend a sweep indefinitely — later orphans are picked up by the next run.

File: pyadcs_connector/PyADCSConnector/tests_discovery.py (hunk 4 above)
Comment (3keyroman) [reply pyadcs_connector/PyADCSConnector/tests_discovery.py]: Addressed in 5400a11. Added `test_sweep_skips_while_a_discovery_holds_the_shared_lock`: it holds `pg_advisory_xact_lock_shared(LOCK_KEY)` on a second connection, asserts `run_cleanup_once()` returns None, deletes nothing and leaves the lease untouched, then releases the lock and asserts the sweep runs and deletes. Also added `test_exclusive_lock_is_released_between_batches`, which probes from a second connection at every batch boundary that a competing shared lock is grantable (verified the probe returns False while an exclusive lock is held, so the assertion is discriminating), and `test_sweep_is_bounded_to_certificates_present_at_start` for the churn/termination bound.

File: pyadcs_connector/PyADCSConnector/services/certificate_cleanup.py (hunk 5 above)
Comment (3keyroman) [reply pyadcs_connector/PyADCSConnector/services/certificate_cleanup.py:46]: Addressed in 5400a11. The lease row is now selected with `select_for_update().order_by("pk").first()`, so the choice is deterministic (lowest pk) even if a second row ever appeared, instead of relying on unordered `.first()`.

File: pyadcs_connector/PyADCSConnector/utils/certificate_fingerprint.py (hunk 6 above)
Comment (3keyroman) [reply pyadcs_connector/PyADCSConnector/utils/certificate_fingerprint.py:4]: Addressed in 5400a11 by documenting it as an explicit accepted limitation in `certificate_fingerprint()`: the exact base64 string is hashed with no decoding or normalization, so two byte-different encodings of the same certificate are stored separately. That is deliberate — it keeps the Python helper and migration 0002 SQL byte-identical and cannot fail on malformed base64 — and costs nothing in practice because ADCS returns each certificate in a stable encoding.

File: pyadcs_connector/pyadcs_connector/wsgi.py (hunk 7 above)
Comment (3keyroman) [reply pyadcs_connector/pyadcs_connector/wsgi.py:34]: Addressed in 5400a11 by documenting both cases next to the entrypoint call in `wsgi.py` (and in the README): `manage.py runserver` also imports the WSGI module and therefore starts the sweep, which `CERTIFICATE_CLEANUP_ENABLED=false` turns off for local development; and gunicorn must not use `--preload`, because the application is then imported by the pre-fork master, so the thread starts there and is not inherited by the forked workers, moving ownership of the sweep outside the worker lifecycle. `entry.sh` deliberately does not use `--preload`.

File: pyadcs_connector/PyADCSConnector/tests_discovery.py (hunk 8 above)
Comment (3keyroman) [reply pyadcs_connector/PyADCSConnector/tests_discovery.py]: Addressed in 5400a11. All four docstrings now describe the behavior directly with no task numbers; a repo-wide scan confirms no remaining plan/task references (or TODO/FIXME markers) in the source.

**Review by lubomirw — APPROVED** [review]
